src/VisualStudio/Core/Def/Implementation/UnusedReferences/Dialog/UnusedReferencesTableProvider.DataSource.cs

```diff
@@ -0,0 +1,151 @@
+
+            public string? DisplayName => null;
+
+            private ImmutableList<SinkManager> _managers = ImmutableList<SinkManager>.Empty;
+            private ImmutableArray<UnusedReferencesEntry> _currentEntries;
+
+            public IDisposable Subscribe(ITableDataSink sink)
+            {
+                return new SinkManager(this, sink);
+            }
+
+            public void ReplaceData(Project project, ImmutableArray<ReferenceUpdate> referenceUpdates)
+            {
+                _currentEntries = referenceUpdates.Select(update => new UnusedReferencesEntry(project, update)).ToImmutableArray();
+                var oldManagers = Volatile.Read(ref _managers);
+                for (int i = 0; (i < oldManagers.Count); ++i)
+                {
+                    oldManagers[i].Sink.AddEntries(_currentEntries);
+                }
+            }
+
+            internal void AddSinkManager(SinkManager manager)
+            {
+                var oldManagers = Volatile.Read(ref _managers);
+                while (true)
+                {
+                    var newManagers = oldManagers.Add(manager);
+                    var results = Interlocked.CompareExchange(ref _managers, newManagers, oldManagers);
+                    if (results == oldManagers)
+                        return;
+
+                    oldManagers = results;
+                }
+            }
+
+            internal void RemoveSinkManager(SinkManager manager)
+            {
+                var oldManagers = Volatile.Read(ref _managers);
+                while (true)
+                {
+                    var newManagers = oldManagers.Remove(manager);
+                    var results = Interlocked.CompareExchange(ref _managers, newManagers, oldManagers);
+                    if (results == oldManagers)
+                        return;
+
+                    oldManagers = results;
+                }
+            }
+
+            internal sealed class SinkManager : IDisposable
+            {
+                internal readonly UnusedReferencesDataSource UnusedReferencesDataSource;
+                internal readonly ITableDataSink Sink;
+
+                internal SinkManager(UnusedReferencesDataSource unusedReferencesDataSource, ITableDataSink sink)
+                {
+                    UnusedReferencesDataSource = Requires.NotNull(unusedReferencesDataSource, nameof(unusedReferencesDataSource));
+                    Sink = sink;
+
+                    UnusedReferencesDataSource.AddSinkManager(this);
+                }
+
+                public void Dispose()
+                {
+                    // Called when the person who subscribed to the data source disposes of the cookie (== this object) they were given.
+                    UnusedReferencesDataSource.RemoveSinkManager(this);
+                }
+            }
+
+            internal class UnusedReferencesEntry : ITableEntry
+            {
+                public Project Project { get; }
```

Comment:
Are these entries being cleared out when we close the dialog? Or are we going to memory leak the Project?

Reply:
Moved to keeping the values I need on the entry instead of the project itself.